site/_posts/2018-12-12-reinventing-the-proxy.md

```diff
@@ -0,0 +1,124 @@
+
+With the help of my colleague [Keenan](https://github.com/kbrock) we came up with the idea of a bouncing select. I was investing a lot of hours to come up with a solution, so I was really shocked when I realized that it can be explained really easily. The arrays passed to the `IO.select` should be dynamic: if a socket becomes ready for reading, it should be temporarily removed (bounced) from the first array. This forces the thread to sleep in any next iteration until a new readiness event happens. After a successful transmission between a pair of sockets, each of them should be added (bounced) back to the corresponding arrays.
+
+```ruby
+loop do
+  read, write, _ = IO.select(@to_read, @to_write, [], 0.1)
+  @to_read -= read
+  @to_write -= write
+
+  # Iterate through the socket pairs that are ready for transmission
+  each_ready do |rd, wr|
+    data = rd.read_nonblock(4096)
+    wr.write_nonblock(translate(data))
+
+    @to_read.push(rd)
+    @to_write.push(wr)
+  end
+end
+```
+
+The performance of this solution is acceptable even with just having a few connections. As the `IO.select` needs to pass the arrays of sockets to the kernel every time it is called, the performance can drop significantly with a higher number of connections. Also the bouncing operations on the arrays add some extra overhead. These problems can be avoided by using `epoll` that uses an internal structure in the kernel space to hold the sockets we are interested in. An application can register sockets to an *epoll instance* and then in a separate call wait for events to happen. Another advantage is that by using `EPOLLONESHOT` it is possible to register a socket for a single readiness event only. This way only half of the bouncing needs to be implemented, the other half happens automatically.
+
+There are multiple gems available that use `epoll` under the hood, unfortunately, none of them implement the aforementioned `EPOLLONESHOT` feature. This is how I ended up writing the proxy as a native extension in my favorite language, the one named after the third letter of the alphabet. The only problem with this solution is that is supported by Linux only. Ideally we should not care about other platforms as we build our appliances with CentOS, but some of our engineers use a Mac for development. There is a very similar solution implemented in BSD-like systems called `kqueue` and it could solve this problem. Unfortunately, it requires a lot of `malloc` calls which can increase the risk of memory leaks. The final solution was to fall back to the `IO.select` on platforms that do not have `epoll` available.
```

Comment:
"is that is supported by Linux only." -> is that **it is** supported by Linux only.